sage: 8.9 -> 9.2

collares opened the pull request:
###### Motivation for this change

@omasanori did most of the work in this update (see #101447), but I had some free time to finish up the work over the last days. All tests pass, except for the fact that `src/sage/interfaces/singular.py` is failing intermittently upstream (https://trac.sagemath.org/ticket/30945).

Keep in mind while reviewing is that I changed SAGE_LOCAL in sage-env.nix. There's a comment on the file explaining the motivation, but I am not sure if I made the right call.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

omasanori committed:
sage: 8.9 -> 9.2

collares committed:
giac: 1.5.0-21 -> 1.5.0-87

collares committed:
gmpy2: 2.1a4 -> 2.1.0b5

collares committed:
ecl: patches to fix SIGFPEs and for unicode compatibility with maxima

collares committed:
maxima-ecl: unpin ecl version

collares committed:
pynac: import patch to avoid deprecation warnings on python3.8

collares committed:
sage: 8.9 -> 9.2 fixups

SuperSandro2000 commented on pkgs/applications/science/math/sage/sage-with-env.nix:
```diff
@@ -13,7 +13,7 @@
 , gmp
 , gfan
-, python2
+, python
```
```suggestion
, python3
```

collares replied:
"python" here is not pkgs.python, because default.nix does
```
  python = pkgs.python3.override {
    packageOverrides = self: super: {
      sagelib = self.callPackage ./sagelib.nix {
...
  }}
  sage-with-env = callPackage ./sage-with-env.nix {
    inherit python pythonEnv;
```
I felt it would be more consistent to use this overriden Python 3 package everywhere instead of using pkgs.python3. Do you think it would be less confusing to rename default.nix's `python` variable to `python3` and pass that along instead?

omasanori replied:
> Do you think it would be less confusing to rename default.nix's python variable to python3 and pass that along instead?

Maybe, as searching `/python[^3]/` is the most easy-to-come-up-with way when one want to find targets for the Python 3 migration.

SuperSandro2000 commented on pkgs/development/compilers/ecl/default.nix:
```diff
@@ -36,6 +36,17 @@ stdenv.mkDerivation {
   };
 
   patches = [
+    # https://git.sagemath.org/sage.git/tree/build/pkgs/ecl/patches/ECL_WITH_LISP_FPE.patch?id=f82c716fdf9c6e91a07166d36b6329a15ecfb41d
```
Can you please use fetchurl?

SuperSandro2000 replied:
Oh derp. You can use fetchurl but fetchpatch is better because it normalizes the patch that changing meta data does not cause a hash mismatch. Sorry about my derp.

collares replied:
Don't worry! I thought about that (it was done like this before), but the patches are plain files committed to Sage's repository, they are not dynamically generated. Does fetchpatch still provide an advantage in this case?

collares commented:
`ecl.pyx` is also intermittently segfaulting on the CI build (it passed before I pushed sagedoc changes which have nothing to do with ECL). A trac search returns nothing, and I can't reproduce the failure locally.

jcumming commented:
I just gave this a whirl, and it works for me. I'll try it in a few more spots and see if I can isolate a failure.

Synthetica9 commented:
I'm still running tests, but the build seems to work. Kudos!

symphorien commented:
I reviewed the giac/xcas update and it looks good to me.

collares commented:
Context: `ecl.py` crashes at

```

Here's the stacktrace for the `ecl.pyx` failure (running under Python instead of IPython, but that shouldn't make a difference). 

```
#0  0x00007f58ee108c71 in sigsegv_handler ()
   from /nix/store/9wzdzij7qwv27ak6p93shmp22xwvx2ha-ecl-20.4.24/lib/libecl.so.20.4
#1  <signal handler called>
#2  0x00007f58ee11aca6 in ecl_cons ()
   from /nix/store/9wzdzij7qwv27ak6p93shmp22xwvx2ha-ecl-20.4.24/lib/libecl.so.20.4
#3  0x00007f58ee0bffb0 in ecl_interpret ()
   from /nix/store/9wzdzij7qwv27ak6p93shmp22xwvx2ha-ecl-20.4.24/lib/libecl.so.20.4
#4  0x00007f58ee0bd5b5 in ecl_apply_from_stack_frame ()
   from /nix/store/9wzdzij7qwv27ak6p93shmp22xwvx2ha-ecl-20.4.24/lib/libecl.so.20.4
#5  0x00007f58ee0bda55 in cl_apply ()
   from /nix/store/9wzdzij7qwv27ak6p93shmp22xwvx2ha-ecl-20.4.24/lib/libecl.so.20.4
#6  0x00007f58ee3337b4 in safe_cl_apply ()
   from /nix/store/cczgihp4z3f20mvgfw74kymjs0avxvsv-python3-3.8.6-env/lib/python3.8/site-packages/sage/libs/ecl.cpython-38-x86_64-linux-gnu.so
#7  0x00007f58ee33bcca in __pyx_f_4sage_4libs_3ecl_ecl_safe_apply ()
   from /nix/store/cczgihp4z3f20mvgfw74kymjs0avxvsv-python3-3.8.6-env/lib/python3.8/site-packages/sage/libs/ecl.cpython-38-x86_64-linux-gnu.so
#8  0x00007f58ee33c213 in __pyx_pw_4sage_4libs_3ecl_9EclObject_15__call__ ()
   from /nix/store/cczgihp4z3f20mvgfw74kymjs0avxvsv-python3-3.8.6-env/lib/python3.8/site-packages/sage/libs/ecl.cpython-38-x86_64-linux-gnu.so
#9  0x00007f5a89bffd47 in _PyObject_MakeTpCall ()
   from /nix/store/fzgk230riagc58n8w1fa41ngs723pzpb-python3-3.8.6/lib/libpython3.8.so.1.0
#10 0x00007f5a89c6d839 in _PyEval_EvalFrameDefault ()
   from /nix/store/fzgk230riagc58n8w1fa41ngs723pzpb-python3-3.8.6/lib/libpython3.8.so.1.0
#11 0x00007f5a89c681f1 in _PyEval_EvalCodeWithName ()
   from /nix/store/fzgk230riagc58n8w1fa41ngs723pzpb-python3-3.8.6/lib/libpython3.8.so.1.0
#12 0x00007f5a89c67ef3 in PyEval_EvalCodeEx ()
   from /nix/store/fzgk230riagc58n8w1fa41ngs723pzpb-python3-3.8.6/lib/libpython3.8.so.1.0
#13 0x00007f5a89cdea5b in PyEval_EvalCode ()
   from /nix/store/fzgk230riagc58n8w1fa41ngs723pzpb-python3-3.8.6/lib/libpython3.8.so.1.0
#14 0x00007f5a89cef718 in run_eval_code_obj ()
   from /nix/store/fzgk230riagc58n8w1fa41ngs723pzpb-python3-3.8.6/lib/libpython3.8.so.1.0
#15 0x00007f5a89cef6b3 in run_mod ()
   from /nix/store/fzgk230riagc58n8w1fa41ngs723pzpb-python3-3.8.6/lib/libpython3.8.so.1.0
#16 0x00007f5a89bbb148 in PyRun_FileExFlags ()
   from /nix/store/fzgk230riagc58n8w1fa41ngs723pzpb-python3-3.8.6/lib/libpython3.8.so.1.0
#17 0x00007f5a89bbaf1b in PyRun_SimpleFileExFlags ()
   from /nix/store/fzgk230riagc58n8w1fa41ngs723pzpb-python3-3.8.6/lib/libpython3.8.so.1.0
#18 0x00007f5a89cf75e6 in Py_RunMain ()
   from /nix/store/fzgk230riagc58n8w1fa41ngs723pzpb-python3-3.8.6/lib/libpython3.8.so.1.0
#19 0x00007f5a89cf6f59 in Py_BytesMain ()
   from /nix/store/fzgk230riagc58n8w1fa41ngs723pzpb-python3-3.8.6/lib/libpython3.8.so.1.0
#20 0x00007f5a896fedbd in __libc_start_main ()
   from /nix/store/a3syww9igm49zdzq3ibzw9m8ccvsgxla-glibc-2.32/lib/libc.so.6
#21 0x000000000040107a in _start ()
```

collares commented:
`ecl/default.nix` does not use boehmgc by default, but `ecl/16.1.2.nix` does. Is that intentional? The intermittent failures in `ecl.pyx` are related to that, and I fixed them by overriding `useBoehmgc` in `sage/default.nix`. I wonder if we should change the default GC, though.

jcumming commented:
I've come across a few more warts:
 - the `nauty` hash changed (upstream modified the tarball with a bugfix)
 - the `pari 2.11.4` tarball moved
 - the `ppl` git repo at https://www.cs.unipr.it/git/gitweb.cgi?p=ppl/ppl.git;a=summary has become inaccessible.

collares committed:
nauty: update tarball sha256

The old `nauty` tarball is currently accessible at
https://distfiles.macports.org/nauty/nauty27r1.tar.gz. The diff is a
single line in genbg.c:

-    SUMMARY(&nout,t2-t1);
+    SUMMARY(nout,t2-t1);

collares committed:
pari: update tarball location, since it moved

collares committed:
ppl: fetch clang patch from alternate location (that doesn't require auth)

Synthetica9 reviewed (approved):
Compiles and runs for me!

Synthetica9 commented:
What's holding this up?

jcumming commented:
My derpy RS code stuff works. Thanks everyone! I support merging this.

omasanori reviewed (approved):
I had reviewed patches and found no problems. (In fact, it is definitely better than the previous iteration.)
Thank you @collares for completing fixes and thank you all for feedbacks!

symphorien commented:
when running nix-review:
```
builder for '/nix/store/n6n8pvxb1417zq6n6y13ngp8vh8z3fyk-python3.8-cypari2-2.1.2.drv' failed with exit code 2; last 10 log lines:
  installing
  
  Usage:   
    pip install [options] <requirement specifier> [package-index-options] ...
    pip install [options] -r <requirements file> [package-index-options] ...
    pip install [options] [-e] <vcs project url> ...
    pip install [options] [-e] <local project path> ...
    pip install [options] <archive url/path> ...
  
  no such option: --build
cannot build derivation '/nix/store/q7xijhksb5lbb36wyb6m306smcxslcn7-python3.8-sagelib-9.2.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/yfdhsgxzk16fyjip5fdz513vbmci3dk4-python3-3.8.6-env.drv': 2 dependencies couldn't be built
cannot build derivation '/nix/store/zynzncb0k79kzgqnmbw2f9b8sry4jdv7-sage-env.drv': 2 dependencies couldn't be built
cannot build derivation '/nix/store/rzcbcifk7zjk680z3mdw3w2v6i6vf8wh-sage-with-env-9.2.drv': 2 dependencies couldn't be built
cannot build derivation '/nix/store/j1phkq8agnyyaxir1d39hrvvykry92jf-jupyter-kernels.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/cf97h6gfpaiqq525zdygws17024m9wxv-sage-tests-9.2.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/f7d2xn5s9mwwbkxjbwg9s2sxqy6l07id-sagedoc-9.2.drv': 2 dependencies couldn't be built
cannot build derivation '/nix/store/a3dg580nclfjsg86crjjq8x1xvcqqg0q-sage-9.2.drv': 4 dependencies couldn't be built
cannot build derivation '/nix/store/xrcsav943264vg7f03nnb4dn5lfs28ay-sage-9.2.drv': 3 dependencies couldn't be built
cannot build derivation '/nix/store/3f9xmdiyakzmjr7wnfafvc5d5qd076s5-env.drv': 2 dependencies couldn't be built
[69 built (1 failed), 873 copied (7375.6 MiB), 1628.0 MiB DL]
error: build of '/nix/store/3f9xmdiyakzmjr7wnfafvc5d5qd076s5-env.drv' failed
https://github.com/NixOS/nixpkgs/pull/105615
2 packages failed to build:
sage sageWithDoc
```

collares committed:
cypari2: revert 2.1.1 -> 2.1.2 update, sync with pip-install-hook.sh

pip install deprecated the --build flag. The standard python
installPhase (in pip-install-hook.sh) got updated in commit
76966f8dddf9b0b5bb2d21128c5fd3e4cf3d6256, but we use a custom one so
we need to update it separately.

collares commented:
I have rebased and reverted the cypari2 update that landed after my PR was submitted (in #105368), since it was decided in #103810 that we would wait for upstream to update cypari2 first (https://trac.sagemath.org/ticket/31029). Hopefully that helps with the failure.

collares commented:
Since there are several people on the Sage maintainer team and @omasanori already reviewed the patches, hopefully asking this will help coordinate: Is reviewing this a pending task on someone else's to-do list? If so, I am very happy to wait to get comments, especially on some potentially questionable decisions I made to get the testsuite to pass. I completely understand that time is a scarce resource, and this message is not intended to rush anyone! If you do intend to review this but haven't gotten around to do it yet, please don't re-order your to-do list because of this comment :)

On the other hand, if no one has a pending review, then I propose we merge this PR, since the tests currently pass (modulo the intermittent `singular.pyx` failure, which also affects upstream and is being investigated there) and there is user interest in getting Sage back into a usable state. There are definitely some parts that need feedback and cleanup, even if just removing "is this OK?" comments I left, but I promise to address any comments or suggestions quickly in a follow-up PR. 

cc @timokau @7c6f434c

symphorien commented:
I ran nix-review successfully on this afternoon (before your last commit). Could you add an explanation in comment to the FIXME you added ?

collares commented:
@symphorien Good catch, thanks! Those lines weren't needed after all, so I just removed them.

timokau commented:
Yes, I had planned to review this. Unfortunately my spare time for open source development is *very* limited recently and this is quite the diff, so I kept putting it off.

I agree with you: This is a definite improvement over the status quo and shouldn't be held off for too long. Sage and nixpkgs moves on, and things might break again. Given your promise to fix up any issues that come up in an "after the fact" review, I'm happy to merge this now and look at it in some more detail later. I have skimmed the changes and nothing sticks out.

Could you start discussions on the relevant sections of the diff where you need additional feedback?

Thank you @collares and @omasanori for your incredible work. I am very, very glad that we found not only one, but two people who have the motivation and skill to do this after I have neglected the package for so long.

collares reviewed (commented):
Many thanks @timokau for all your hard work over the years on packaging Sage, and @omasanori for doing most of the work in getting this back to a working state! Sorry for the delay in submitting this, I thought I had done it yesterday but I forgot to click "Submit review" :)

yurrriq commented:
Thanks, everyone! Especially @omasanori!

omasanori commented on pkgs/applications/science/math/sage/sagedoc.nix:
```diff
@@ -47,13 +44,26 @@ stdenv.mkDerivation rec {
     chmod -R 755 "$SAGE_DOC_SRC_OVERRIDE"
   '';
 
+  postPatch = ''
```
I saw some similar *patch represented as a sequence of commands* `postPatch` instances, so it is fine IMHO.
